### ISSUE TITLE
Support library-specific plugin

### ISSUE THREAD

**Issue by ikatyang** [prettier/prettier#4424]
from https://github.com/prettier/prettier/pull/4423#pullrequestreview-117811698:

> I think that we should start to figure out a way to do those kind of library-specific kind of transforms into plugins.
> 
> There are a lot of downsides with the current approach of bundling it inside of prettier:
> - Now we need to maintain all those ourselves
> - If jest changes their format, then you need to wait a new prettier release to be able to use it. This dramatically slow down iteration speed
> - Ideally jest would be in control of how their format their templates
> - We have a good relationship with jest but what if some library coming from someone else wanted to format their thing, we now have to figure out what specific library printing is worth being inside of prettier.

**Comment by ikatyang** [comment]
I think we can have a `printerPlugin` entry in a plugin:

```ts
// some-plugin.js
module.exports = {
  printerPlugin: {
    // astFormat
    estree: (path, print, textToDoc, options) => Doc | null;
  }
}
```

which should work something like [`embed()`](https://prettier.io/docs/en/plugins#printers), i.e. potentially switch to a different printerPlugin in every node:

https://github.com/prettier/prettier/blob/1a6cf3d0717c5828d9bf474d13d115e3150ee254/src/main/ast-to-doc.js#L83-L99

**Comment by suchipi** [comment]
I'm super in favor of this; it would also solve https://github.com/prettier/prettier/issues/4260 and let users create plugins on top of prettier for the places we don't want to add options

**Comment by j-f1** [comment]
For example, someone could make a `JSXAttribute` plugin that prints string values with single quotes.

**Comment by j-f1** [comment]
One thing we’ll have to figure out is what to do if two plugins try to customize the printing of a node.

**Comment by suchipi** [comment]
Hyperscript and mithril would be good candidates for llibrary plugins; see https://github.com/prettier/prettier/issues/4424

**Comment by j-f1** [comment]
@suchipi looks like that’s this issue 😕

**Comment by bakkot** [comment]
Drive-by comment - while I think something like this would be a good idea, especially for things like formatting tagged template literals, I'm worried that a really general approach to plugins will mean that people will just start implementing configuration as plugins, which I think would eliminate a lot of [value to end users](https://github.com/prettier/prettier/issues/40) of prettier's conservative approach to options: i.e., people would start fighting about configuration again, just now in the form of "add this third-party plugin to format this really specific thing the way I want it to be formatted" instead of "add this config option to [etc]".

Maybe that won't happen, since the bar for adding another dev-dependency is often much higher than the bar for adding config, for many projects, but I thought it worth mentioning.

Side note: this will also arguably mean that any changes to prettier's internal doc format would require major version bumps.

**Comment by azz** [comment]
> I'm worried that a really general approach to plugins will mean that people will just start implementing configuration as plugins

This was my reaction, too. Having these kinds of plugins will create fragmentation.

**Comment by j-f1** [comment]
We could limit it only to tagged template strings, allowing folks to create custom formatting for template strings.

**Comment by suchipi** [comment]
For hyperscript and functional programming, users might want to change how call expressions, function expressions, and function declarations are printed. But I'm not sure what a clean API for that would look like.

**Comment by SimenB** [comment]
Could it be useful for stuff like #2630, or is that something else again?

**Comment by suchipi** [comment]
It might work as a workaround, but I'd like decorators to work out-of-the-box for users of either version of the plugin eventually, if we can support that.

**Comment by suchipi** [comment]
@bakkot I like the proposed interface, but I think we should roll it as part of our plugin API. The main export of a plugin module is an object with a `languages` property, a `parsers` property, etc. We could also support a `templateTagFormatters` property on the main export of a plugin. See https://prettier.io/docs/en/plugins.html#developing-plugins

Also, I think something worth considering here is that there is a lot of overlap between formatting tagged template strings and formatted fenced code blocks in markdown. If there is a way to support both, that would be cool- maybe just a `fencedCodeBlockFormatters` which works similarly

**Comment by meyer** [comment]
If only formatting template tag literals was supported, you all would be missing out on a big opportunity IMO. I realise that a very powerful plugin API lets users do things you don’t want them to do, but it also enables people to do weird and wonderful things with prettier. There’s only so much hand-holding you can do before you have to let go and let users do their own thing.

I found this issue because I’m looking to format the values of JSX attributes on a set of components from a known source. I know what I want to do in regards to AST modification, but I can’t find the Prettier Approved way of doing it. I would need full access to the AST as I would be visiting each JSXElement and checking to see if it’s from the desired source before modifying it.

I’m not sure exactly how things are done around here so I don’t really have any useful advice to offer, but the initial proposal is pretty close to what I’m looking for. I think the visitor model would work quite well here.

**Comment by duailibe** [comment]
@meyer If all you want is to modify the AST you can use the [custom parser API](https://prettier.io/docs/en/api.html#custom-parser-api) to change what you want and let the Prettier do the printing afterwards.

**Comment by meyer** [comment]
@duailibe fantastic, thanks!

**Comment by suchipi** [comment]
@meyer it would, assuming JSXElement is in the whitelist (which seems to make sense). Out of curiosity, what would you want your plugin to do?

**Comment by meyer** [comment]
@suchipi I’m trying to run the values of string props on certain JSX components through prettier. I only want to prettify props that come from my library. Example:

```jsx
import { Block } from 'jsxstyle';

// before
<Block color="rgba(0,0,0,.2)" />

// after ✨
<Block color="rgba(0, 0, 0, 0.2)" />
```

I’ve got a demo working on my machine. The last big problem for me to solve is getting the source of the module somehow. I can whitelist the [known set of component names](https://github.com/jsxstyle/jsxstyle/blob/27cdf74/packages/jsxstyle-utils/src/componentStyles.ts#L2-L17) (`Block`, `Inline`, `Box`, etc.), but they’re generic enough that collisions are highly likely.

**Comment by meyer** [comment]
@j-f1 yeah, I’ve implemented the required scope magic to identify the source module in a [different thing](https://github.com/jsxstyle/jsxstyle/tree/master/packages/jsxstyle-webpack-plugin). I was originally attempting to plug it in using multiparser functionality, but in my next attempt I’m going to re-parse (😩) and and modify the AST directly. Since I’m modifying a string value it can (and probably should) happen before prettier does its thing.

**Comment by j-f1** [comment]
You might be able to put into Prettier by making a new module that calls Prettier now passes [a custom `parser` option when appropriate](https://prettier.io/docs/en/api.html#custom-parser-api). @azz has created [a module](https://github.com/azz/prettier-transform/blob/master/README.md) that makes this much easier.

**Comment by meyer** [comment]
@j-f1 that’s perfect, thanks! (sorry for the rabbit trail)

**Comment by toomuchdesign** [comment]
I'd like to write a small Prettier plugin/transformer targeting `JSX` to give [this issue](https://github.com/prettier/prettier/issues/3101) a try.

I know very little about Prettier's internals, I made a quick attempt and I'm honestly a bit lost. Is there any resource or significative part of Prettier's source code I might read to find some inspiration?

Thank you!

**Comment by j-f1** [comment]
@toomuchdesign I don’t think we support plugins that change how an already supported language is formatted right now, unfortunately.

**Comment by toomuchdesign** [comment]
I see @j-f1, I probably underestimated the challenge! I'll keep following the discussion, though. Thanks for your efforts, here! 👍

**Comment by suchipi** [comment]
Maybe we should start with an internal-use-only API that lets you do a lot, so that we can use it to convert current library-specific formatting rules to plugins (internally), so that we can see what kind of things those type of plugins need, so we can reduce the API's scope before allowing public use?

**Comment by jquense** [comment]
It seems like the current plugin API does support this? Or is there a complexity I am missing in the API?

**Comment by thorn0** [comment]
@jquense This is still the case: https://github.com/prettier/prettier/issues/4424#issuecomment-408489742

But it's probably possible to achieve something by monkey-patching Prettier's printers. See https://github.com/prettier/prettier/issues/8195

**Comment by Acidic9** [comment]
Are there any updates to this? The thread started in 2018 but seems like it hasn't been implemented yet?

**Comment by void-m4110c** [comment]
Any news?

**Comment by ivank** [comment]
@duailibe @bakkot What If instead of adding additional properties to the plugin api, prettier figured out which language were inside embedded nodes and attempted to find a plugin for it?

My current predicament is that I'm writing a parser / prettifier for SQL inside js template literals.

```typescript
import sql from '@potygen/query';

const consumptionByTopic = sql`
  SELECT
    id,
    topic
  FROM consumptions
  WHERE topic = $topic
  LIMIT 1
`;
```

I have the parsing, introspection and type generation, and prettier plugin implemented, but I can't seem to be able to tell prettier to run my plugin inside of TS template literal.

I'm sure I'm not alone in building a DSL in template literals, or prettifying existing one. I'm already using a VSCode plugin to add syntax highlighting for the SQL itself, and would be building a code completion plugin soon.

You could have a heuristic to figure out the "language" of a template literal or other embed "ports" and use appropriate plugin to format what's inside. You're already doing it with official prettier plugins, why not allow this to be used by others as well. I'd imagine you'll use it only if it doesn't match anything you have already implemented.

I'll be happy to attempt to implement it as a PR myself if you guys think it is a good approach?

**Comment by karlhorky** [comment]
@ivank this sounds super interesting for us and our bootcamp students (we use and teach [Postgres.js](https://github.com/porsager/postgres)).

Is there anywhere else I can watch the progress on these things you mentioned?

**Comment by ivank** [comment]
The progress of it can be found here: https://github.com/ivank/potygen - its currently built to support our own work - generating types and prettifying SQL. We don't currently support _tagged_ template literals but that's on our radar too.

Similar amount of features where we'd like to arrive at could be seen with [ts-mysql-plugin](https://github.com/segmentio/ts-mysql-plugin) though that's for MYSQL. The amount of work they've put into this I'm sure they'd have the ability and desire to write a prettifier if it was possible.

**Comment by ivank** [comment]
Just noticed this in prettier's official xml plugin, where if there are parsers with a given name, just use them - can something like this work, at least for some specific use cases?

https://github.com/prettier/plugin-xml/blob/b0749d546ff4a96d9125ae25d57509c794c01171/src/embed.ts#L87-L99

Maybe that's  way forward?

**Comment by csantos1113** [comment]
I think this 👇 is somewhat related to this issue,

I wanted to extend the prettier-html functionality to replace some characters to other characters, for example “” to ", so I tried and failed.
Then  I saw https://github.com/prettier/prettier/issues/10423#issuecomment-787140819 and https://github.com/prettier/prettier/issues/10259#issuecomment-774785489 comments from @thorn0

> _The plugin API is for adding parsers/printers, not for patching existing ones. There are hacky ways to do that, but that is not officially supported and can break in any release. See https://github.com/prettier/prettier/issues/10259_

And I was 🙃😢.

But then I kept testing with custom prettier plugins and find out that this is possible

```js
module.exports = {
  parsers: {
    html: {
      astFormat: 'html',
      parse(text, parsersParam, options) {
        console.log('Yep, parser replaced and I broke prettier!');
        return { hello: 'world' };
      }
    }
  }
};
```

So I guess my question/comment is why I can't provide custom `printers` to extend HTML (or any other) but I can provide custom `parsers` that break the whole thing 😞

**Comment by viveleroi** [comment]
I've found this issue after following a trail of several. I was hoping to use prettier on a major rewrite of a new app but it's making changes to our code that I didn't ask for in the config. For example, it's turning an array with entries on multiple lines into a single line.

```json
{
  "recommendations": [
    "arcanis.vscode-zipfs",
    "esbenp.prettier-vscode",
    "EditorConfig.EditorConfig"
  ]
}
```

turns into:

```json
{ 
  "recommendations": ["arcanis.vscode-zipfs", "esbenp.prettier-vscode", "EditorConfig.EditorConfig"]
}
```

There's a significant amount of conversation to read but the gist is that: a) users asked for a way to disable this and b) the devs weren't interested in adding rules. Three years later it appears, if I'm reading this all correctly, there's still no solution outside of adding ignores all over the place.

Why are there any forced formatting changes to begin with? Unless there's a fix for this I'm missing, we have to drop prettier entirely because we can't allow tools that make forced changes to our code.

**Comment by danriedl** [comment]
@viveleroi check out this plugin: [electrovir/prettier-plugin-multiline-arrays](https://github.com/electrovir/prettier-plugin-multiline-arrays). right now you it doesn't work with v3 [but the repo owner is working on it](https://github.com/electrovir/prettier-plugin-multiline-arrays/issues/26#issuecomment-1646964000)

**Comment by Sec-ant** [comment]
I'm working on a plugin called [`prettier-plugin-embed`](https://github.com/Sec-ant/prettier-plugin-embed) which can be seen as an effort to abstract out the embedded language formatting logic off Prettier and make it configurable and pluggable. It was developed to mainly address the problem mentioned in #5588 and my personal needs, with the following features:

- It supports some additional languages like XML, SQL, GLSL and new ones can be added gradually with relatively little effort.
- It provides options to disable formatting for certain languages, even those ones from Prettier.
- It provides options to customize the tag identifiers you want to use for detecting certain languages.

 I would be happy if others find this plugin helpful.

**Comment by karlhorky** [comment]
@Sec-ant thanks so much for this!

I can confirm, this works well with formatting SQL inside tagged template literals with [`prettier-plugin-sql`](https://github.com/un-ts/prettier/tree/master/packages/sql) by @JounQin 🙌 

- https://github.com/Sec-ant/prettier-plugin-embed/issues/20

**Comment by AlexRMU** [comment]
So what?

**Comment by AlexRMU** [comment]
Why can't you just run all the plugins defined for the language one by one in turn?

**Comment by quantuminformation** [comment]
was language injection supported yet?

would love to run some magic on this

![image](https://github.com/prettier/prettier/assets/216566/e76fd74c-79a8-4d68-ba49-e8a280e73553)

**Comment by JounQin** [comment]
@quantuminformation https://github.com/Sec-ant/prettier-plugin-embed is the thing you want.

**Comment by karlhorky** [comment]
Since it may not be clearly surfaced from [the humble comment](https://github.com/prettier/prettier/issues/4424#issuecomment-1780954068) above, formatting of SQL-in-JS (and other embedded languages like XML-in-JS, YAML-in-JS, PHP-in-JS, Ruby-in-JS, Markdown-in-JS, and more) is possible with @Sec-ant's  highly-recommended [`prettier-plugin-embed` plugin](https://github.com/Sec-ant/prettier-plugin-embed)

For example, I tweeted on X/Twitter (https://x.com/karlhorky/status/1736365230303715701) about formatting of SQL-in-JS:

https://github.com/prettier/prettier/assets/1935696/fe0f93e1-d6a3-454f-8658-505fedbc4d7e

**Comment by Harpush** [comment]
I also need a way to extend html printing process... It would be awesome if I could create a plugin that just alters a specific area printing and reuses the regular html printing everywhere else. I can do it with private API but that's not ideal.
